### PR TITLE
Added Micronaut SnapStart version (#12)

### PR DESCRIPTION
**Pull request by dmahapatro**
*Issue #, if available:*
#12 

*Description of changes:*
Added SnapStart sample for Micronaut

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: micronaut/pom.xml [hunk 1]
```diff
@@ -45,6 +45,11 @@
       <artifactId>micronaut-function-aws</artifactId>
       <scope>compile</scope>
     </dependency>
+    <dependency>
+      <groupId>io.micronaut.crac</groupId>
+      <artifactId>micronaut-crac</artifactId>
+      <version>1.2.0</version>
```
Comment (alvarosanchez) [inline micronaut/pom.xml:51]: Micronaut BOM v3.8.7 brings 1.1.2. If you really need 1.2.0, I'd recommend omitting the version here and setting the following property:

```xml
<micronaut.crac.version>1.2.0</micronaut.crac.version>
```

It should make future version upgrades easier.

Comment (dmahapatro) [reply micronaut/pom.xml:51]: Good callout @alvarosanchez . I was looking for the latest version.

Comment (alvarosanchez) [reply micronaut/pom.xml]: I'll add @timyates and @yawkat to the loop as our CRaC and benchmarking experts, respectively :)

**Comment by dmahapatro** [comment]
Thanks @yawkat and @alvarosanchez. Looking at the PR do you see any other areas of improvement other than Eager Initialization of Singletons and SnapStart in general.

File: micronaut/src/main/java/software/amazonaws/example/product/PrimingResource.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+@Singleton
+public class PrimingResource implements OrderedResource {
+  private static final Logger log = LoggerFactory.getLogger(PrimingResource.class);
+  private final DynamoProductDao dynamoProductDao;
+
+  public PrimingResource(DynamoProductDao dynamoProductDao) {
+    this.dynamoProductDao = dynamoProductDao;
+  }
+
+  public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
```
Comment (ravsrin) [inline micronaut/src/main/java/software/amazonaws/example/product/PrimingResource.java:20]: Do you need to override the beforeCheckpoint() and afterCheckpoint() methods? Similar to Spring Boot pattern and in this Micronaut [blog](https://micronaut.io/2022/11/28/leveraging-aws-lambda-snapstart-with-the-micronaut-framework/).

File: micronaut/src/main/java/software/amazonaws/example/product/PrimingResource.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.slf4j.LoggerFactory;
+import software.amazonaws.example.product.dao.DynamoProductDao;
+
+@Singleton
+public class PrimingResource implements OrderedResource {
+  private static final Logger log = LoggerFactory.getLogger(PrimingResource.class);
+  private final DynamoProductDao dynamoProductDao;
+
+  public PrimingResource(DynamoProductDao dynamoProductDao) {
+    this.dynamoProductDao = dynamoProductDao;
```
Comment (ravsrin) [inline micronaut/src/main/java/software/amazonaws/example/product/PrimingResource.java:17]: Spring Boot pattern sets the `Core.getGlobalContext().register(this);`, is this something we need here too?

**Review by ravsrin — COMMENTED** [review]
Provided some feedback. Deployed stack successfully and ran Artillery test to confirm benchmarks. Only ran it once and Cold and Warm start numbers were similar.

File: README.md [hunk 4]
```diff
@@ -136,6 +136,17 @@ It fits particularly well with Lambda to reduce the initialization time, but doe
             <th scope="col">p99</th>
             <th scope="col">max</th>
         </tr>
+        <tr>
+            <th>Micronaut</th>
```
[the remark this answers is not in the document]
Comment (dmahapatro) [reply README.md:140]: Nothing major. I ran the other 2 scenarios, the numbers were pretty much similar.

File: micronaut/src/main/java/software/amazonaws/example/product/PrimingResource.java (hunk 2 above)
Comment (dmahapatro) [reply micronaut/src/main/java/software/amazonaws/example/product/PrimingResource.java:20]: `micronaut-crac` does that under the hood.

File: micronaut/src/main/java/software/amazonaws/example/product/PrimingResource.java (hunk 3 above)
Comment (dmahapatro) [reply micronaut/src/main/java/software/amazonaws/example/product/PrimingResource.java:17]: `micronaut-crac` does that under the hood.

File: micronaut/src/main/java/software/amazonaws/example/product/handler/BaseProductHandler.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package software.amazonaws.example.product.handler;
```
[the remark this answers is not in the document]
Comment (dmahapatro) [reply micronaut/src/main/java/software/amazonaws/example/product/handler/BaseProductHandler.java:4]: Good call. Modified.

File: micronaut/src/main/java/software/amazonaws/example/product/PrimingResource.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package software.amazonaws.example.product;
```
[the remark this answers is not in the document]
Comment (dmahapatro) [reply micronaut/src/main/java/software/amazonaws/example/product/PrimingResource.java:4]: Done.

File: micronaut/src/main/java/software/amazonaws/example/product/handler/BaseProductHandler.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * Refer <a href="https://docs.micronaut.io/latest/guide/#eagerInit">Eager Init of Singletons</a> for details
+ */
+@Introspected
+public abstract class BaseProductHandler extends MicronautRequestHandler<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {
+  @Nonnull
+  @Override
+  protected ApplicationContextBuilder newApplicationContextBuilder() {
+    ApplicationContextBuilder builder = super.newApplicationContextBuilder();
+    builder.eagerInitSingletons(true);
```
[the remark this answers is not in the document]
Comment (dmahapatro) [reply micronaut/src/main/java/software/amazonaws/example/product/handler/BaseProductHandler.java:22]: I had the same train of thoughts while writing the BaseProductHandler

**Comment by ravsrin** [comment]
Looks good. Nice job!

**Review by muraria — APPROVED** [review]
Looks Good. Thanks.

**Pull request merged**